feat: add support of all tx types in the history screen

OlegMakarenko opened the pull request:
![image](https://user-images.githubusercontent.com/33131259/149224017-2895e6ab-e890-4fca-a5f1-bd084fe9b380.png)

**What the current behaviour is**
- 20 transaction types are not supported by the mobile wallet history screen.
- Signer and recipient addresses are not presented in transaction details.

**Why this is an issue**
- Wallet should support all transaction types.
- Wallet should present signer and recipient addresses and other data related to the transaction in the details.

**What's the fix**
- Added support of all tx types in the history screen.
- Added inner tx preview for an aggregate tx.
- Removed usless methods from FetchTransaction service.
- Recreated TransactionItem component.
- Extended transaction details. Resolved: #90 
- Fixed the mosaic info extra API call. Related: #280 
- Added unit tests.
- Replace duplicated code with utils fns.

![image](https://user-images.githubusercontent.com/33131259/149224282-50f7d3bc-4822-48c8-be85-b342d7d7c021.png)

cryptoBeliever commented:
1. For unlink and link key is the same icon. We should use the strike-through key icon for UNLINK?
2. For non-aggregate transactions for Unconfirmed there is a date presented instead "Unconfirmed" label. 
3. I think it would be very helpful to present some icon that is telling if the row is an outgoing or incoming transaction.

OlegMakarenko commented:
@cryptoBeliever 

> For unlink and link key is the same icon. We should use the strike-through key icon for UNLINK?
Added new icon and 'link' caption to the transaction type.

> For non-aggregate transactions for Unconfirmed there is a date presented instead "Unconfirmed" label.
Fixed.

> I think it would be very helpful to present some icon that is telling if the row is an outgoing or incoming transaction.
We have different icons for incoming and outgoing transfer tx, or do you mean something different?

cryptoBeliever commented:
@OlegMakarenko 

> > For unlink and link key is the same icon. We should use the strike-through key icon for UNLINK?
> 
> Added new icon and 'link' caption to the transaction type.

Thank you. New icon is presented now :+1: 

> 
> > For non-aggregate transactions for Unconfirmed there is a date presented instead "Unconfirmed" label.

Works correctly :+1: 

> 
> > I think it would be very helpful to present some icon that is telling if the row is an outgoing or incoming transaction.
> 
> We have different icons for incoming and outgoing transfer tx, or do you mean something different?

Yes, please ignore this. I thought about other than transfer transaction types but it doesn't make much sense.

OlegMakarenko commented on src/utils/transaction.js:
```diff
@@ -0,0 +1,192 @@
+/**
+ * Checks whether transaction is awaiting a signature by account.
+ */
+export const transactionAwaitingSignatureByAccount = (transaction, account, multisigAddresses) => {
+    if (transaction.transactionType === TransactionType.AGGREGATE_BONDED) {
+        const accountPublicKey = getPublicKeyFromPrivateKey(account.privateKey);
+        const transactionSignerAddresses = transaction.innerTransactions.map(innerTransaction => innerTransaction.signerAddress);
+        const cosignRequired = transactionSignerAddresses.some(
+            transactionSignerAddress => transactionSignerAddress && multisigAddresses?.some(address => address === transactionSignerAddress)
+        );
+        const transactionHasMissingSignatures = !!transaction.transactionInfo?.merkleComponentHash.startsWith('000000000000');
+        const signedByCurrentAccount = transaction.cosignaturePublicKeys.some(publicKey => publicKey === accountPublicKey);
+
+        return (
+            (!signedByCurrentAccount && transaction.status !== Constants.Message.CONFIRMED) ||
+            (transactionHasMissingSignatures && cosignRequired)
+        );
+    }
+
+    return false;
+};
```
@bassemmagdy moved [this part](https://github.com/symbol/mobile-wallet/pull/230/files#diff-4ca36d50c878e4284c9d5c900368a8b61dc59e90bccf9433355f71b622da80ecL73-R101) here

yilmazbahadir commented on src/utils/transaction.js:
```diff
@@ -0,0 +1,192 @@
+        amount = getMosaicRelativeAmount(currencyMosaic);
+    }
+
+    if (amount && !outgoingTransaction) {
+        infoPreview.push({
+            type: TransactionInfoPreviewValueType.AmountIncoming,
+            value: amount,
+        });
+    }
+    if (amount && outgoingTransaction) {
+        infoPreview.push({
+            type: TransactionInfoPreviewValueType.AmountOutgoing,
+            value: amount,
+        });
+    }
```
Should we refactor this with if-else ?

OlegMakarenko replied:
Thanks, added

yilmazbahadir commented on src/components/controls/Icon.js:
```diff
@@ -241,6 +242,105 @@ export default class C extends Component<Props, State> {
                 source = require('@src/assets/icons/namespace.png');
                 break;
+
+            case `transaction_${TransactionType.TRANSFER}`:
```
This switch statement was too big and it's getting bigger imo. Should we consider replacing it with an associative array instead? Something like this:

```
let src = '';
const iconArray = [];
iconArray[`none`] = '';
iconArray[`transaction_${TransactionType.TRANSFER}`] = '@src/assets/icons/transaction/transfer.png';
iconArray[`transaction_${TransactionType.TRANSFER}_incoming`] ='@src/assets/icons/transaction/incoming.png';
iconArray[`transaction_${TransactionType.TRANSFER}_outgoing`] = '@src/assets/icons/transaction/outgoing.png'
...
// end of iconArray fill
src = iconArray[name] || '@src/assets/icons/ic-about.png';
if (src) {
    source = require(src);
}
// rest of the function
```

OlegMakarenko replied:
Seems it is not possible to do this way.

> the image name in require has to be known statically.

This is from React Native docs (https://reactnative.dev/docs/images#static-image-resources)

```
// GOOD
<Image source={require('./my-icon.png')} />;

// BAD
var icon = this.props.active
  ? 'my-icon-active'
  : 'my-icon-inactive';
<Image source={require('./' + icon + '.png')} />;

// GOOD
var icon = this.props.active
  ? require('./my-icon-active.png')
  : require('./my-icon-inactive.png');
<Image source={icon} />;


Another option is to import them all together. But need to ensure it will not cause any performance issue:

const iconSrcMap = {
    [`transaction_${TransactionType.HASH_LOCK}`]: require('@src/assets/icons/transaction/lock.png'),
    [`transaction_${TransactionType.SECRET_PROOF}`]: require('@src/assets/icons/transaction/cross-chain.png'),
    [`transaction_${TransactionType.VRF_KEY_LINK}`]: require('@src/assets/icons/transaction/key.png'),
    [`transaction_${TransactionType.ACCOUNT_KEY_LINK}`]: require('@src/assets/icons/transaction/key.png'),
    [`transaction_${TransactionType.NODE_KEY_LINK}`]: require('@src/assets/icons/transaction/key.png'),
    [`transaction_${TransactionType.VOTING_KEY_LINK}`]: require('@src/assets/icons/transaction/key.png'),
    ...
}

return <Image source={iconSrcMap[name]} />
```

yilmazbahadir replied:
I see, better to leave it as it is then 👍
